sdks/python/apache_beam/io/gcp/spannerio.py

```diff
@@ -0,0 +1,333 @@
+    self._spanner_configuration = spanner_configuration
+    self._snapshot = None
+
+  def to_runner_api_parameter(self, context):
+    return self.to_runner_api_pickled(context)
+
+  def setup(self):
+    spanner_client = Client(self._spanner_configuration.project)
+    instance = spanner_client.instance(self._spanner_configuration.instance)
+    database = instance.database(self._spanner_configuration.database,
+                                 pool=self._spanner_configuration.pool)
+    self._snapshot = BatchSnapshot.from_dict(database, self._snapshot_dict)
+
+  def process(self, element):
+    with self._snapshot._get_session().transaction() as transaction:
+      for row in getattr(transaction, element.transaction_action)(
+          **element.kwargs):
+        yield row
+
+  def teardown(self):
+    if self._snapshot:
+      self._snapshot.close()
+
+
+class _NaiveSpannerRead(PTransform):
+  """
+  A naive version of Spanner read that use transactions for read and execute
+  sql methods from the previous state.
+  """
+
+  def __init__(self, transaction, read_operations, spanner_configuration):
+    self._transaction = transaction
+    self._read_operations = read_operations
+    self._spanner_configuration = spanner_configuration
+
+  def expand(self, pbegin):
+    return (pbegin
+            | 'Add Read Operations' >> beam.Create(self._read_operations)
+            | 'Reshuffle' >> beam.Reshuffle()
+            | 'Perform Read' >> beam.ParDo(
+                _NaiveSpannerReadDoFn(
+                    snapshot_dict=self._transaction,
+                    spanner_configuration=self._spanner_configuration
+                )))
+
+
+class _BatchRead(PTransform):
+  """
+  This transform uses the Cloud Spanner BatchSnapshot to perform reads from
+  multiple partitions.
+  """
+
+  def __init__(self, read_operations, spanner_configuration):
+
+    if not isinstance(spanner_configuration, _BeamSpannerConfiguration):
+      raise ValueError("spanner_configuration must be a valid "
+                       "_BeamSpannerConfiguration object.")
+
+    self._read_operations = read_operations
+    self._spanner_configuration = spanner_configuration
+
+  def expand(self, pbegin):
+    spanner_client = Client(project=self._spanner_configuration.project,
+                            credentials=self._spanner_configuration.credentials,
+                            user_agent=self._spanner_configuration.user_agent)
+    instance = spanner_client.instance(self._spanner_configuration.instance)
+    database = instance.database(self._spanner_configuration.database,
+                                 pool=self._spanner_configuration.pool)
+    snapshot = database.batch_snapshot(**self._spanner_configuration
+                                       .snapshot_options)
+
+    reads = [
+        {"read_operation": ro.read_operation, "partitions": p}
+        for ro in self._read_operations
+        for p in getattr(snapshot, ro.batch_action)(**ro.kwargs)
```

Comment:
Please make sure that the way you generate partitions here is compatible with the Java source.